File: 17.电话号码的字母组合.py
#
# @lc app=leetcode.cn id=17 lang=python3
#
# [17] 电话号码的字母组合
#

# @lc code=start
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        if len(digits) == 0:
            return res
        num_dict = {'2': ['a', 'b', 'c'],
                    '3': ['d', 'e', 'f'],
                    '4': ['g', 'h', 'i'],
                    '5': ['j', 'k', 'l'],
                    '6': ['m', 'n', 'o'],
                    '7': ['p', 'q', 'r', 's'],
                    '8': ['t', 'u', 'v'],
                    '9': ['w', 'x', 'y', 'z']}
        for i in range(len(digits)):
            if i == 0:
                res = res + num_dict[digits[0]]
            else:
                while len(res[0]) == i:
                    tmp = res.pop(0)
                    for ch in num_dict[digits[i]]:
                        res.append(tmp+ch)
        return res
# @lc code=end

File: 17.电话号码的字母组合.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=17 lang=java
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> combs = new ArrayList<String>();
        if(digits.length()==0){
            return combs;
        }
        Map<Character, String> phone = new HashMap<Character,String>() {
            {
                put('2', "abc");
                put('3', "def");
                put('4', "ghi");
                put('5', "jkl");
                put('6', "mno");
                put('7', "pqrs");
                put('8', "tuv");
                put('9', "wxyz");
            }
        }; 
        back(digits, 0, combs, phone, new StringBuffer());
        return combs;
    }

    public void back(String digits, int index, List<String> combs, Map<Character, String> phone, StringBuffer com){
        if(index == digits.length()){
            combs.add(com.toString());
        }else{
            for(int i=0; i<phone.get(digits.charAt(index)).length(); i++){
                com.append(phone.get(digits.charAt(index)).charAt(i));
                back(digits, index+1, combs, phone, com);
                com.deleteCharAt(index);
            }
        }
    }
}
// @lc code=end

